content/blog/2019-12-18-ml-based-phases-ordering-with-llvm/index.md

```diff
@@ -0,0 +1,75 @@
++++
+title = "ML based phases ordering with LLVM"
+extra.author = "Qian Huang and Horace He"
+extra.bio = """
+Qian Huang is a junior undergraduate studying CS and Mathematics.
+
+Horace He is a senior undergraduate studying CS and Mathematics.
+"""
++++
+
+In this project, we continued to experiment with ML based phases ordering as in project 2, but using c++ programs and LLVM instead. Compared with project 2 using Bril, there are a lot more infrastructure supports for this project: more available optimization passes, actual hardware performance counters and benchmark programs. These allowed us to explore this topic more.
+
+# Design Overview
+
+## Analysis
+
+We first analyzed how much the ordering of optimization passes affects program performances. We selected the 270 optimizations used in LLVM `-O2` level as the set of optimizations we analyzed across the whole project. We performed all optimizations on the streamcluster program in PARSEC benchmark.
+
+![Performances using randomly selected passes](./w.png)
+
+For each n optimization passes n, we uniformly randomly select n optimization passes with replacement (so there can be duplicate paths) and run 10 times to obtain a range of performance. From the figure, we see that the ordering of optimizations does affect the overall program performance significantly. In fact, some order of optimizations even makes it significantly worse. As the number of passes used increases, we can see that the optimal performance improvement indeed increases but saturated around 100 in terms of the best optimization performances. Thus we can find a smaller and better optimization sequence.
```

Comment:
- Perhaps there is a duplicate "n" in "n optimization passes n"? Maybe this is just supposed to say "For each n, …"?
- paths -> passes?
- saturated -> saturates (present tense, to match "increases")
- performance (singular again)